src/libraries/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs

```diff
@@ -1337,4 +1366,166 @@ internal bool CanAddValue(HttpHeaderParser parser)
+                    }
+
+                    Debug.Assert(value is string);
+                    return new HeaderStringValues((string)value);
+                }
+            }
+
+            public KeyValuePair<string, HeaderStringValues> Current { get; private set; }
+
+            public void Dispose()
+            {
+                _headerStore.Dispose();
+            }
+
+            object? IEnumerator.Current => Current;
+
+            void IEnumerator.Reset()
+            {
+                _headerStore.Reset();
+                Current = default;
+            }
+        }
+    }
+
+    public readonly struct HeaderStringValues : IEnumerable<string>
+    {
+        private readonly IEnumerable<string>? _values;
+        private readonly string? _value;
+
+        public HeaderStringValues(string value)
+        {
+            _value = value;
+            _values = null;
+        }
+
+        public HeaderStringValues(IEnumerable<string> value)
+        {
+            _values = value;
+            _value = null;
+        }
+
+        public Enumerator GetEnumerator() => _values != null ? new Enumerator(_values!) : new Enumerator(_value!);
+
+        IEnumerator<string> IEnumerable<string>.GetEnumerator() => GetEnumerator();
+
+        IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+
+        public struct Enumerator : IEnumerator<string>
+        {
+            private readonly IEnumerator<string>? _values;
+            private readonly string? _value;
+            private bool _completed;
+
+            internal Enumerator(IEnumerable<string> values)
+            {
+                _values = values.GetEnumerator();
+                _value = null;
+                _completed = false;
+                Current = default!;
+            }
+
+            internal Enumerator(string value)
+            {
+                _values = null;
+                _value = value;
+                _completed = false;
+                Current = default!;
+            }
+
+            public bool MoveNext()
+            {
+                if (_completed)
+                {
+                    return false;
+                }
+
+                if (_values == null)
+                {
+                    _completed = true;
+                    Current = _value!;
+                    return true;
+                }
+
+                _completed = !_values.MoveNext();
+                if (_completed)
+                {
+                    return false;
+                }
+
+                Debug.Assert(_values.Current != null);
+                Current = _values.Current!;
+
+                return true;
+            }
+
+            public string Current { get; private set; }
+
+            public void Dispose()
+            {
+                _values?.Dispose();
+            }
+
+            object? IEnumerator.Current => Current;
+
+            void IEnumerator.Reset()
+            {
+                _values?.Reset();
```

Comment:
You don't actually need to implement Reset, you can just throw NotImplementedException. It's not used by foreach or LINQ or anything like that.

Reply:
I know about that, but throwing NotImplementedException from IEnumerator.Reset is a bad choice for 3rd party library developers. If you writing some generic code working with IEnumerator, you'd expect it to fully implement contract. I personally ran into this issue in the past.

Reply:
> If you writing some generic code working with IEnumerator, you'd expect it to fully implement contract. I personally ran into this issue in the past.

No. The vast majority of Reset methods throw.  And it's perfectly acceptable to do so, even called out in the docs as such:
"The Reset method is provided for COM interoperability. It does not necessarily need to be implemented; instead, the implementer can simply throw a NotSupportedException."
No one can depend on an arbitrary Reset method working.

Reply:
Ok, then. I will change it to throw NIE.

Reply:
Done